pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -336,11 +298,114 @@ func (a *index) buildGlobalCollections(
 		opts,
 	)
 
-	WorkloadAddressIndex := krt.NewIndex[networkAddress, model.WorkloadInfo](GlobalWorkloads, "networkAddress", networkAddressFromWorkload)
-	WorkloadServiceIndex := krt.NewIndex[string, model.WorkloadInfo](GlobalWorkloads, "service", func(o model.WorkloadInfo) []string {
+	workloadNetworkServiceIndex := krt.NewIndex[string, model.WorkloadInfo](GlobalWorkloads, "network;service", func(o model.WorkloadInfo) []string {
+		res := make([]string, 0, len(o.Workload.Services))
+		for svc := range o.Workload.Services {
+			res = append(res, strings.Join([]string{o.Workload.Network, svc}, ";"))
+		}
+		return res
+	})
+
+	coalescedWorkloads := krt.NewManyCollection(
+		workloadNetworkServiceIndex.AsCollection(),
+		func(ctx krt.HandlerContext, i krt.IndexObject[string, model.WorkloadInfo]) []model.WorkloadInfo {
+			parts := strings.Split(i.Key, ";")
+			if len(parts) != 2 {
+				log.Errorf("Invalid key %s for SplitHorizonWorkloads, expected <network>;<service>", i.Key)
+				return nil
+			}
+			networkID := network.ID(parts[0])
+			localNetwork := a.Network(ctx).String()
+			if networkID.String() == localNetwork {
+				// We don't coalesce workloads for the local network
+				return nil
+			}
+
+			svcName := parts[1]
+			svc := krt.FetchOne(ctx, a.services.Collection, krt.FilterKey(svcName))
```

Comment:
We haven't assigned `a.services.Collection` yet; this could race. I suspect this means we have a circular dependency (i.e. we need these workloads to create the final services). Probably means we need a refactor?

Reply:
We can just use `GlobalWorkloadServicesWithCluster` from above as we don't need the splithorizon view of the services. Good catch.